File: 109_Convert_Sorted_List_to_Binary_Search_Tree/109_1.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        size = 0
        ptr = head
        while ptr:
            size += 1
            ptr = ptr.next

        def convert(l, r):
            nonlocal head
            if l > r:
                return None
            mid = (l + r ) // 2
            left = convert(l, mid - 1)
            node = TreeNode(head.val)
            node.left = left
            head = head.next
            node.right = convert(mid + 1, r)
            return node
        return convert(0, size - 1)

File: 109_Convert_Sorted_List_to_Binary_Search_Tree/109_1.c
#include <stdio.h>
#include <string.h>
#include <math.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

struct ListNode * findMiddleElement(struct ListNode *head) {
    struct ListNode *prevPtr = NULL;
    struct ListNode *slowPtr = head;
    struct ListNode *fastPtr = head;
    while (fastPtr != NULL && fastPtr->next != NULL) {
        prevPtr = slowPtr;
        slowPtr = slowPtr->next;
        fastPtr = fastPtr->next->next;
    }
    if (prevPtr != NULL) prevPtr->next = NULL;
    return slowPtr;
}

struct TreeNode * sortedListToBST(struct ListNode *head){
    if (head == NULL) return NULL;
    struct ListNode *mid = findMiddleElement(head);
    struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    node->val = mid->val;
    node->left = NULL;
    node->right = NULL;
    if (head == mid) return node;
    node->left = sortedListToBST(head);
    node->right = sortedListToBST(mid->next);
    return node;
}

static void traverse(struct TreeNode *tree) {
    printf("%d ", tree->val);
    if (tree->left)  traverse(tree->left);
    if (tree->right) traverse(tree->right);
}

int main(void) {
    struct ListNode *node1 = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *node2 = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *node3 = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *node4 = (struct ListNode *)malloc(sizeof(struct ListNode));
    struct ListNode *node5 = (struct ListNode *)malloc(sizeof(struct ListNode));
    node1->val = -10;
    node2->val = -3;
    node3->val = 0;
    node4->val = 5;
    node5->val = 9;
    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;
    traverse(sortedListToBST(node1));
}

